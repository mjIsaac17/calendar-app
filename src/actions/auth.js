import { fetchNoToken, fetchToken } from "../helpers/fetch";
import { types } from "../types/types";

const baseEndpoint = "auth";

export const startLogin = (email, password) => {
  //the return is triggered by thunk when we need to use async functions
  return async (dispatch) => {
    const resp = await fetchNoToken(baseEndpoint, { email, password }, "POST");
    const body = await resp.json();
    if (resp.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(login({ uid: body.uid, name: body.name }));
      return "";
    }
    // else{
    //     //Using sweet alert
    //     //Swal.fire
    // }
    return body.msg;
  };
};

export const startRegister = (email, password, name) => {
  return async (dispatch) => {
    const resp = await fetchNoToken(
      `${baseEndpoint}/new`,
      { email, password, name },
      "POST"
    );
    const body = await resp.json();
    if (resp.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(login({ uid: body.uid, name: body.name }));
      return "";
    }
    return body.msg;
  };
};

//Renew token
export const startChecking = () => {
  return async (dispatch) => {
    const resp = await fetchToken(`${baseEndpoint}/renew`);
    const body = await resp.json();
    if (resp.ok) {
      localStorage.setItem("token", body.token);
      localStorage.setItem("token-init-date", new Date().getTime());
      dispatch(login({ uid: body.uid, name: body.name }));
    } else dispatch(checkingFinish());
  };
};

const checkingFinish = () => ({ type: types.authCheckingFinish });

const login = (user) => ({
  type: types.authLogin,
  payload: user,
});
