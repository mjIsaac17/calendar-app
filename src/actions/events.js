import { fetchToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepareEvents";
import { types } from "../types/types";

export const eventStartAddNew = (event) => {
  return async (dispatch, getState) => {
    try {
      const { uid, name } = getState().auth;
      const resp = await fetchToken("events", event, "POST");
      const body = await resp.json();
      if (resp.ok) {
        event.id = body.event.id;
        event.user = {
          _id: uid,
          name,
        };
        dispatch(eventAddNew(event));
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event,
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event,
});

export const eventClearActive = () => ({
  type: types.eventClearActive,
});

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {
      const resp = await fetchToken(`events/${event.id}`, event, "PUT");
      const body = await resp.json();
      if (resp.ok) {
        dispatch(eventUpdate(event));
      } else {
        alert(`Error. ${body.msg}`);
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventUpdate = (event) => ({
  type: types.eventUpdate,
  payload: event,
});

export const eventStartDelete = () => {
  return async (dispatch, getState) => {
    try {
      const { id } = getState().calendar.activeEvent;
      const resp = await fetchToken(`events/${id}`, {}, "DELETE");
      const body = await resp.json();
      if (resp.ok) {
        dispatch(eventDelete());
      } else {
        alert(`Error. ${body.msg}`);
      }
    } catch (error) {
      console.log(error);
    }
  };
};

const eventDelete = () => ({
  type: types.eventDelete,
});

export const eventStartLoading = () => {
  return async (dispatch) => {
    try {
      const resp = await fetchToken("events");
      const body = await resp.json();
      const events = prepareEvents(body.events);
      dispatch(eventLoaded(events));
    } catch (error) {
      console.log(error);
    }
  };
};

const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events,
});

export const eventLogout = () => ({
  type: types.eventLogout,
});
