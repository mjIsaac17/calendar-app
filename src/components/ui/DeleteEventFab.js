import React from "react";
import { useDispatch } from "react-redux";
import { eventStartDelete } from "../../actions/events";

export const DeleteEventFab = () => {
  const dispatch = useDispatch();
  const handleDelete = () => {
    dispatch(eventStartDelete());
  };
  return (
    <button className="btn btn-danger fab-danger" onClick={handleDelete}>
      <i className="fa fa-trash mr-1"></i>
      <span>Delete event</span>
    </button>
  );
};
