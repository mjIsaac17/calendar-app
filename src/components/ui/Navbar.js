import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Redirect } from "react-router-dom";
import { logout } from "../../actions/auth";
import { eventLogout } from "../../actions/events";

export const Navbar = () => {
  const { name } = useSelector((state) => state.auth);
  const dispatch = useDispatch();

  const handleLogout = () => {
    dispatch(logout());
    dispatch(eventLogout());
    return <Redirect to="/login" />;
  };

  return (
    <div className="navbar navbar-dark bg-dark mb-4">
      <span className="navbar-brand">{name}</span>
      <button className="btn btn-outline-danger" onClick={handleLogout}>
        <i className="fa fa-sign-out-alt mr-1"></i>
        <span>Exit</span>
      </button>
    </div>
  );
};
