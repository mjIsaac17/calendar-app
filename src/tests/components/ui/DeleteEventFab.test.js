import thunk from "redux-thunk";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { DeleteEventFab } from "../../../components/ui/DeleteEventFab";
import { eventStartDelete } from "../../../actions/events";

jest.mock("../../../actions/events", () => ({
  eventStartDelete: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};
const store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
  <Provider store={store}>
    <DeleteEventFab />
  </Provider>
);

describe("Tests in <DeleteEventFab />", () => {
  test("should load the component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should call the eventStartDelete when clicking", () => {
    wrapper.find("button").prop("onClick")();
    expect(eventStartDelete).toHaveBeenCalled();
  });
});
