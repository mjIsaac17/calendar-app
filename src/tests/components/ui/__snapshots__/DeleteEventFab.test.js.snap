// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <DeleteEventFab /> should load the component 1`] = `
<button
  className="btn btn-danger fab-danger"
  onClick={[Function]}
>
  <i
    className="fa fa-trash mr-1"
  />
  <span>
    Delete event
  </span>
</button>
`;
